### PR TITLE
Add capability to send Custom Events to SignalFx.

### PR DESCRIPTION
**Pull request by mmusik**
Please note I will update the README in a separate PR, since we have a big changes there to be merged with next release and I want to work on a new version of the README to avoid a merge.

Changes:
1) Expose API to send custom events to SignalFx
2) Add a dedicated helper to send Cloudwatch Events to SignalFx
3) Increase minor version (api changes).

Open questions/ things I would appreciate insight on:
1) is the choice of properties vs dimensions ok?
2) are we aware of any limitations when it comes to the size of the event?

### CODE REVIEW

File: signalfx-transform-helper.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+function isPrimitive(x) {
+  return Object(x) !== x;
+}
+
+function sanitize(name) {
+  return name.replace(/[^a-zA-Z0-9\-_]+/gi, '_').substring(0, MAX_FIELDNAME_LENGTH);
+}
+
+function flatten(obj, newObj, prefix) {
+  for (let entry of Object.entries(obj)) {
```
Comment (bgola-signalfx) [inline signalfx-transform-helper.js]: `for (let [key, value] of Object.entries(obj)) {`

File: signalfx-transform-helper.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+const FIELD_SEPARATOR = '_';
+const MAX_FIELDNAME_LENGTH = 128;
+
+function isPrimitive(x) {
+  return Object(x) !== x;
+}
+
+function sanitize(name) {
+  return name.replace(/[^a-zA-Z0-9\-_]+/gi, '_').substring(0, MAX_FIELDNAME_LENGTH);
```
Comment (bgola-signalfx) [inline signalfx-transform-helper.js]: Maybe we should log a warning when sanitized name is longer than the `MAX_FIELDNAME_LENGTH`?

File: spec/signalfx-transform-helper-spec.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+const sinon = require('sinon');
+const chai = require('chai');
+const expect = chai.expect;
+
+describe('signalfx-transform-helper', () => {
+
+  it('should leave flat objects as they are', () => {
+    const obj = { 'key1': 'value', 'key2': 5};
+    const flattened = signalfxHelper.toKeyValueMap(obj);
+    expect(flattened.key1).to.be.equal('value');
```
Comment (bgola-signalfx) [inline spec/signalfx-transform-helper-spec.js]: `expect(flattened).to.deep.equal(obj)`

File: signalfx-helper.js [hunk 4]
```diff
@@ -36,22 +51,52 @@ function sendMetric(metricName, metricType, metricValue, dimensions={}) {
...
+function sendCustomizedEvent(type, dimensions, properties, timestamp) {
+  let event = {
+    category: 'USER_DEFINED',
+    eventType: type,
+    dimensions: dimensions,
+    properties: properties
+  };
+
+  if (timestamp) {
+    event = Object.assign(event, {timestamp: timestamp});
```
[the remark this answers is not in the document]
Comment (mmusik) [reply signalfx-helper.js]: done.

File: signalfx-helper.js [hunk 5]
```diff
@@ -23,9 +26,21 @@ if (!isNaN(timeoutMs)) {
   CLIENT_OPTIONS.timeout = 300;
 }
 
-var defaultDimensions, metricSender;
+const clearSendPromises = () => {
```
[the remark this answers is not in the document]
Comment (mmusik) [reply signalfx-helper.js]: done

File: signalfx-helper.js [hunk 6]
```diff
@@ -36,22 +51,52 @@ function sendMetric(metricName, metricType, metricValue, dimensions={}) {
...
+    details = toKeyValueMap({detail: cwEvent.detail});
+    resources = toKeyValueMap({resources: cwEvent.resources});
+  } catch (err) {
+    console.error('Unable to convert details or resources to a key value map. They wont be included in the event.', err);
+  }
+
+  let sfxEvent = {
+    category: 'USER_DEFINED',
+    eventType: CLOUDWATCH_EVENT_TYPE,
+    dimensions: {region: cwEvent.region, account: cwEvent.account, 'detail-type': cwEvent['detail-type'], source: cwEvent.source},
```
[the remark this answers is not in the document]
Comment (mmusik) [reply signalfx-helper.js]: done

File: spec/signalfx-transform-helper-spec.js (hunk 3 above)
Comment (mmusik) [reply spec/signalfx-transform-helper-spec.js]: done

File: signalfx-transform-helper.js (hunk 2 above)
Comment (mmusik) [reply signalfx-transform-helper.js]: done

**Pull request merged**